renderdoc/driver/metal/metal_resources.h

```diff
@@ -0,0 +1,61 @@
+/******************************************************************************
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2022 Baldur Karlsson
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#pragma once
+
+#include "core/resource_manager.h"
+#include "metal_types.h"
+
+struct WrappedMTLObject
+{
+  WrappedMTLObject() = delete;
+  WrappedMTLObject(void *mtlObject, ResourceId objId, WrappedMTLDevice *wrappedMTLDevice)
+      : wrappedObjC(NULL), real(mtlObject), id(objId), m_WrappedMTLDevice(wrappedMTLDevice)
+  {
+  }
+  virtual ~WrappedMTLObject() = default;
```

Comment:
In the future, the code for object lifetime is currently implemented like this

In the wrapped objc object which is tied to the lifetime of the real objc object
```
- (void)dealloc
{
  self.wrappedCPP->Dealloc();
  [super dealloc];
}

```
then
```
void WrappedMTLObject::Dealloc()
{
  GetResourceManager()->ReleaseWrappedResource(this);
}
```
and 
```
template <typename realtype>
  void ReleaseWrappedResource(realtype obj, bool clearID = false)
  {
  ....
  ....
  ....
      delete obj;
  }
  ```

I will remove the virtual now and see if it can be tackled differently later ie. adding a Dealloc() method to each class derived from `WrappedMTLObject`